File: SPEECH/Speech/language/src/main/java/com/javadude/speech/ItemHolder.java
package com.javadude.speech;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class ItemHolder extends Item {
	private Map<String, Item> items = new HashMap<String, Item>();

	public ItemHolder(String name, String description, boolean canBePickedUp) {
		super(name, description, canBePickedUp);
	}
	
	public String dumpContents() {
		String article = "The ";
		if ("inventory".equals(getName()))
			article = "Your ";
		if (isEmpty())
			return article + getName() + " is empty";
		String result = article + getName() + " contains ";
		for (String item : items.keySet()) {
			result += "a " + item + ", ";
		}
		result = result.substring(0, result.length()-2);
		return result;
	}
	public boolean isEmpty() {
		return items.isEmpty();
	}
	public void add(Item item) {
		items.put(item.getName(), item);
	}
	public Item remove(String name) {
		return items.remove(name);
	}
	public Item get(String name) {
		return items.get(name);
	}
	public boolean contains(String name) {
		return items.containsKey(name);
	}
	public Set<String> getItemNames() {
		return items.keySet();
	}
}

File: Module - Speech/Speech2/language/src/main/java/com/javadude/speech2/ItemHolder.kt
package com.javadude.speech2

open class ItemHolder(
                name: String,
                description: String,
                canBePickedUp: Boolean)
        : Item(name, description, canBePickedUp) {

    private val items = mutableMapOf<String, Item>()
    val isEmpty: Boolean
        get() = items.isEmpty()
    val itemNames: Set<String>
        get() = items.keys

    fun dumpContents(): String {
        var article = "The "
        if ("inventory" == name)
            article = "Your "
        if (isEmpty)
            return "$article$name is empty"
        var result = "$article$name contains "
        for (item in items.keys) {
            result += "a $item, "
        }
        return result.dropLast(2)
    }

    fun add(item: Item?) {
        item?.let {
            items[it.name] = it
        }
    }

    fun remove(name: String): Item? {
        return items.remove(name)
    }

    operator fun get(name: String): Item? {
        return items[name]
    }

    operator fun contains(name: String): Boolean {
        return items.containsKey(name)
    }
}
